### PR TITLE
load balancing: Disabling zone aware routing for non-zero priority levels.

### PR DESCRIPTION
**Pull request by alyssawilk**
Restructuring the LB class for impending weighted failover.
Disabling zone aware routing for non-zero priority levels.

*Risk Level*: Medium
*Testing*: Added a unit test for the new function.  It'll be extended in the next PR.
*Docs Changes*: https://github.com/envoyproxy/data-plane-api/pull/359  
*Release Notes*: priorities already in relnotes
Continued work on #1929

### CODE REVIEW

**Comment by alyssawilk** [comment]
@mattklein123 I think this one needs your input, when you have a chance.

**Comment by mattklein123** [comment]
@alyssawilk functionally this seems fine to me, but I'm a little confused about the ultimate goal. Can you potentially rephrase the description into discrete options that we can choose from and then describe the one you recommend and why? (It's mostly there and maybe I'm being tired/dense but am having trouble fully parsing, I think just a slight expansion would help a lot).

**Comment by alyssawilk** [comment]
So we have to figure out how locality and priority interact.

We can (as you suggested in prior PRs) look at all the localities across priorities, percentages of anticipated load, and calculate ideal routing based on that.  IMO that seems complicated and (I think?) unnecessary given I don't anticipate failovers in the local locality.

We can do best-effort per-priority locality determinations.  So the same calculations we have been doing, ignoring the percentage of traffic spilling across priorities.  This is what we have here and (I think) would work well enough if someone decides to have local failovers.

We can do no locality routing at P>0.  This would enforce my belief it would silly to have local failovers (if you have local capacity, use it!  But someone could add support later) and also be a fairly simple and straight-forward and probably cleaner to explain.  Now that I think about it this may be a better choice just because it's more straight-forward to document.  I'll switch to this if you agree with me :-)

This is complicated by me not really grokking locality based routing.  I think I'm too stuck in a world with GSLB where you have smarts in the sky delivering you optimal traffic flows and you don't have to think about this sort of thing :-P

**Comment by alyssawilk** [comment]
On more thought I think I prefer to leave as-is.  Depending on the granularity of locality I could see the 'local' locality being at the DC layer and lb-endpoints in different clusters, so if a full cluster were drained, locality based routing for the failover would make sense.

**Comment by mattklein123** [comment]
I'm going to be honest in saying that this entire thing is giving me a headache which I think is an indication that it might be too complicated. What do you think about just skipping locality routing at P > 0 as you originally proposed until someone complains about it? If you think that's a bad idea I think we should probably have a short meeting about this as I think it would better to talk through in real time.

**Comment by alyssawilk** [comment]
Haha, works for me.  Disabled and documented.  Someone who cares can implement in the distant future :-P

File: source/common/upstream/load_balancer_impl.cc [hunk 1]
```diff
@@ -79,9 +84,11 @@ ZoneAwareLoadBalancerBase::ZoneAwareLoadBalancerBase(const PrioritySet& priority
                const std::vector<HostSharedPtr>&) -> void {
           ASSERT(priority == 0);
           UNREFERENCED_PARAMETER(priority);
-          // If the set of local Envoys changes, regenerate routing based on potential changes to
-          // the set of servers routing to priority_set_.
-          regenerateLocalityRoutingStructures(bestAvailablePriority());
+          // If the set of local Envoys changes, regenerate routing for all priority levels based on
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: Do we actually need this change if we are only doing locality at P=0 now?

**Comment by alyssawilk** [comment]
Sure, we can skip a few operations and not wait for earlyExitNonLocalityRouting.

Aside, with the current structure of Envoy code, are we guaranteed an updateHosts?
I'd assume the LB is created with the initial host set, and it looks like we never generate locality routing structures for the initial host set (only on modification).  This predates all my priority changes so if it is actually a bug I'll fix in a follow-up PR, probably in the new year.

**Comment by mattklein123** [comment]
> Aside, with the current structure of Envoy code, are we guaranteed an updateHosts?
I'd assume the LB is created with the initial host set, and it looks like we never generate locality routing structures for the initial host set (only on modification). This predates all my priority changes so if it is actually a bug I'll fix in a follow-up PR, probably in the new year.

I would have to check, but IIRC if there are no hosts at all, everything is short circuited to return nullptr so it doesn't matter if there are no locality structures.

**Comment by alyssawilk** [comment]
Fair enough.  I think I'll still add a test in a follow up.
Anything else you want on this one?

**Comment by alyssawilk** [comment]
Oh ugh, last upload didn't upload.  Will push once I'm back by my laptop :-(

**Pull request merged**